src/core/execute.c

```diff
@@ -5696,44 +5711,88 @@ int exec_spawn(Unit *unit,
+
+        fdset = fdset_new();
+        if (!fdset)
+                return log_oom();
 
-                _exit(exit_status);
+        r = unit_serialize(unit, f, fdset, /* switching_root */ false);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize unit '%s': %m", unit->id);
+
+        r = exec_command_serialize(command, f);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecCommand: %m");
+
+        r = exec_parameters_serialize(params, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecParameters: %m");
+
+        r = exec_runtime_serialize(runtime, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecRuntime: %m");
+
+        r = exec_context_serialize(context, f, fdset);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize ExecContext: %m");
+
+        r = exec_cgroup_context_serialize(cgroup_context, f);
+        if (r < 0)
+                return log_unit_error_errno(unit, r, "Failed to serialize CGroupContext: %m");
+
+        if (fseeko(f, 0, SEEK_SET) == (off_t) -1)
+                return log_unit_error_errno(unit, errno, "Failed to reseek on serialization stream: %m");
+
+        r = fd_cloexec(fileno(f), false);
+        if (r < 0)
+                return log_unit_error_errno(unit, errno, "Failed to set O_CLOEXEC on serialization fd: %m");
+
+        r = fdset_cloexec(fdset, false);
+        if (r < 0)
+                return log_unit_error_errno(unit, errno, "Failed to set O_CLOEXEC on serialized fds: %m");
+
+        /* The executor binary is pinned, to avoid compatibility problems during upgrades. */
+        xsprintf(cb_arg.serialization_fd_number, "%i", fileno(f));
+        cb_arg.executor_fd = unit->manager->executor_fd;
+        cb_arg.executor_path = FORMAT_PROC_FD_PATH(unit->manager->executor_fd);
+
+        if (stack_size == 0) {
+                struct rlimit rl;
+
+                r = getrlimit(RLIMIT_STACK, &rl);
+                if (r < 0)
+                        return log_unit_error_errno(unit, errno, "Failed to get RLIMIT_STACK: %m");
+
+                stack_size = MIN(rl.rlim_cur, page_size());
         }
 
+        stack = mmap(NULL,
+                     stack_size,
+                     PROT_READ | PROT_WRITE,
+                     MAP_PRIVATE | MAP_ANONYMOUS | MAP_STACK,
+                     /* fd= */ -1,
+                     /* offset= */ 0);
+        if (stack == MAP_FAILED)
+                return log_unit_error_errno(unit, errno, "Failed to allocate stack: %m");
+
+        pid = clone(clone_cb, stack + stack_size, CLONE_VFORK|CLONE_VM|SIGCHLD, &cb_arg);
+        if (pid < 0)
+                return log_unit_error_errno(unit, errno, "Failed to clone: %m");
```

Comment:
please move the mmap() + clone() into a helper call of its own.

hmm, once you clone() like this, the child will have its own stack separate from ours, but it won't be a copy of it, but start ad empty state, no?

accessing local variables shouldn't work then anymore? should it?

Reply:
> hmm, once you clone() like this, the child will have its own stack separate from ours, but it won't be a copy of it, but start ad empty state, no?
>
> accessing local variables shouldn't work then anymore? should it?

That's what I thought at the beginning, but it is working, so... apparently it does